# Trading Day Of the Week (TDOW)

Vamos a analizar cuantitativamente la serie historica de precios del **MERVAL** (5 años aprox) para observar **la probabilidad de que el cierre diario del activo sea al alza o baja segun el dia de la semana.** Esto nos va a servir como un indicador para tomar en cuenta al momento de operar, y tambien va a servir para explicar que los mercados no se comportan como **random walk** debido a que estan predispuesto a cerrar arriba o abajo segun el tipo del dia. Aparte de calcular la probabilidad, vamos a calcular **el promedio del rendimiento segun cada dia de la semana.** Se tomaron datos de Octubre 1996 a septiembre 2018.

### Importamos los paquetes y librerias necesarios para el analisis

In [1]:
import pandas as pd
import numpy as np

### Cambio el directorio a la carpeta "Datasets" donde voy a tener el archivo de los datos historicos

In [2]:
cd C:\Users\david\Documents\Datasets

C:\Users\david\Documents\Datasets


### Corroboro que la direccion presente de trabajo sea igual al directorio cambiado, donde estan las series

In [3]:
pwd

'C:\\Users\\david\\Documents\\Datasets'

### Importo el CSV y lo agrego a un dataframe llamado "data"

In [4]:
data = pd.read_csv('MERV.csv',low_memory=False)

### Uso .head() para visualizar las 10 primeras filas del dataframe

In [5]:
data.head()

Date        Open        High         Low       Close   Adj Close  \
0  1996-10-08  579.429993  592.650024  579.429993  590.109985  590.109985   
1  1996-10-09  590.109985  590.830017  582.210022  583.140015  583.140015   
2  1996-10-10  583.239990  584.609985  578.270020  584.609985  584.609985   
3  1996-10-11  584.909973  585.869995  582.440002  584.320007  584.320007   
4  1996-10-14  584.289978  584.289978  580.609985  584.030029  584.030029   

   Volume  
0     0.0  
1     0.0  
2     0.0  
3     0.0  
4     0.0

### Observo la forma de la serie (filas x columnas)

In [6]:
data.shape

(5536, 7)

### Busco si faltan valores, True=Si False=No

Como aparece todo False (No), comprobamos que en este otro sentido, la serie esta "limpia".

In [7]:
data.isnull().any()

Date         False
Open          True
High          True
Low           True
Close         True
Adj Close     True
Volume        True
dtype: bool

### Como tiene valores que faltan, comparo y veo si sacando los datos erroneos, cambia substancialmente las observiones

In [8]:
len(data), len (data.dropna())

(5536, 5371)

### Cambio los valores de "data" por la serie limpia

In [9]:
data = data.dropna()

### Corroboro que la cant de observaciones cambio

In [10]:
len(data)

5371

### Cambio el formato de los datos de la columna "Date" al de fecha mediante pd.to_datetime

In [11]:
data['Date'] = pd.to_datetime(data['Date'])

### Una vez que esta en formato datetime, puedo cambiar las fechas segun dia de semana

Utilizo .head() para ver las primeras 5 filas segun el formato del dia de semana

In [12]:
data['Date'].dt.weekday.head()

0    1
1    2
2    3
3    4
4    0
Name: Date, dtype: int64

### Creo un diccionario para asignarle un string value al formato numerico del dia de semana

In [13]:
day_dict = {
    0:'Lunes',
    1:'Martes',
    2:'Miercoles',
    3:'Jueves',
    4:'Viernes',
    5:'Sabado',
    6:'Domingo'
           }

### Agrego una columna para asignarle el dia de semana a cada fecha, utilizando el diccionario que creamos antes

In [14]:
data['Day_of_week']=data['Date'].dt.weekday.map(day_dict)

In [15]:
data.head()

Date        Open        High         Low       Close   Adj Close  \
0 1996-10-08  579.429993  592.650024  579.429993  590.109985  590.109985   
1 1996-10-09  590.109985  590.830017  582.210022  583.140015  583.140015   
2 1996-10-10  583.239990  584.609985  578.270020  584.609985  584.609985   
3 1996-10-11  584.909973  585.869995  582.440002  584.320007  584.320007   
4 1996-10-14  584.289978  584.289978  580.609985  584.030029  584.030029   

   Volume Day_of_week  
0     0.0      Martes  
1     0.0   Miercoles  
2     0.0      Jueves  
3     0.0     Viernes  
4     0.0       Lunes

### Creo una columna para calcular el PnL diario

Calculamos la ganancia que obtendriamos al comprar a la apertura y vender al cierre

In [16]:
data['PnL'] = data['Close'] - data['Open']

### Creo una columna para referirme si el PnL fue negativo o postivio
0 = Negativo
1 = Positivo

In [17]:
data['Pos_neg'] = data['PnL'].apply(lambda x: int(x >=0))

In [18]:
data.head()

Date        Open        High         Low       Close   Adj Close  \
0 1996-10-08  579.429993  592.650024  579.429993  590.109985  590.109985   
1 1996-10-09  590.109985  590.830017  582.210022  583.140015  583.140015   
2 1996-10-10  583.239990  584.609985  578.270020  584.609985  584.609985   
3 1996-10-11  584.909973  585.869995  582.440002  584.320007  584.320007   
4 1996-10-14  584.289978  584.289978  580.609985  584.030029  584.030029   

   Volume Day_of_week        PnL  Pos_neg  
0     0.0      Martes  10.679992        1  
1     0.0   Miercoles  -6.969970        0  
2     0.0      Jueves   1.369995        1  
3     0.0     Viernes  -0.589966        0  
4     0.0       Lunes  -0.259949        0

### Creo una columna para calcular los retornos logaritmicos del PnL (comprar a la apertura y vender al cierre)

In [19]:
data['Log_Ret_PnL'] =  np.log(data['Close']/data['Open'])
data.head()

Date        Open        High         Low       Close   Adj Close  \
0 1996-10-08  579.429993  592.650024  579.429993  590.109985  590.109985   
1 1996-10-09  590.109985  590.830017  582.210022  583.140015  583.140015   
2 1996-10-10  583.239990  584.609985  578.270020  584.609985  584.609985   
3 1996-10-11  584.909973  585.869995  582.440002  584.320007  584.320007   
4 1996-10-14  584.289978  584.289978  580.609985  584.030029  584.030029   

   Volume Day_of_week        PnL  Pos_neg  Log_Ret_PnL  
0     0.0      Martes  10.679992        1     0.018264  
1     0.0   Miercoles  -6.969970        0    -0.011882  
2     0.0      Jueves   1.369995        1     0.002346  
3     0.0     Viernes  -0.589966        0    -0.001009  
4     0.0       Lunes  -0.259949        0    -0.000445

### Sumamos los Pos_neg segun el dia de semana
Esto nos va a permitir saber cuantos dias son los positivos, sobre el total de observaciones y la probabilidad que el PnL (Open-Close) sea positiva

In [20]:
group1 = data.groupby(['Day_of_week']).agg(['sum','size'])['Pos_neg']
group1['Prob_PnL +'] = group1['sum']/group1['size']
group1

sum  size  Prob_PnL +
Day_of_week                       
Jueves       559  1083    0.516159
Lunes        510  1006    0.506958
Martes       574  1097    0.523245
Miercoles    607  1109    0.547340
Viernes      568  1076    0.527881

**Los dias mas propensos a cerrar al alza son los Miercoles con un 54.73% de probabilidad.**

### Agrupamos segun dia de semana y segun si el PnL fue negativo o positivo

Ademas de eso, calculamos la media de los retornos logaritmicos segun el Pos_neg

In [21]:
data.groupby(['Day_of_week','Pos_neg']).agg(['mean'])[['Log_Ret_PnL']]

Log_Ret_PnL
                           mean
Day_of_week Pos_neg            
Jueves      0         -0.015096
            1          0.014509
Lunes       0         -0.017457
            1          0.014790
Martes      0         -0.014895
            1          0.013964
Miercoles   0         -0.014650
            1          0.013552
Viernes     0         -0.012905
            1          0.013727

**Como se analizo previamente, el miercoles es el dia mas inclinado historicamente a cerrar al alza. El mismo tiene un promedio de rendimiento de 1.47% en estos dias de PnL positivos.**